### DISCUSSION TITLE
What is wrong with my join function?

### DISCUSSION THREAD

**Discussion by realapril** [simolus3/drift#1546] [Q&A]
I followed [this instruction](https://drift.simonbinder.eu/docs/advanced-features/joins/) and I have a problem here
```
return query.watch().map((rows) {
  return rows.map((row) {
    return EntryWithCategory(
      row.readTable(todos),
      row.readTableOrNull(categories),
    );
  }).toList();
});
```

problematic lines are these,
```
row.readTable(todos),
row.readTableOrNull(categories),
```
the error message says 
```
The argument type '$ToDoTable' can't be assigned to the parameter type 'ResultSetImplementation<$ToDoTable, ToDo>'. 
```

What have I done wrong? How should I change it?
Acctually my longer code is this one
```
Stream<List<EntryWithCategory>> entriesWithCategory() =>
  select(todos).join([
    leftOuterJoin(categories, categories.id.equalsExp(todos.category)),
  ])
 .watch()
.map((rows) => rows
 .map((row) => (EntryWithCategory(
      row.readTable(todos),
      row.readTableOrNull(categories),
    ))).toList());

```

I tried with this one too but it made the same error
```
Stream<List<EntryWithCategory>> entriesWithCategory() {
    final query = select(todos).join([
      leftOuterJoin(categories, categories.id.equalsExp(todos.category)),
    ]);
    return query.watch().map((rows) {
      return rows.map((row) {
        return EntryWithCategory(
          row.readTable(todos),
          row.readTableOrNull(categories),
        );
      }).toList();
    });
  }
```

**Comment by simolus3** [comment]
Could you post the versions of `drift` and `drift_dev` that you're using? The generated `$ToDoTable` should mix in `TableInfo` which in turn implements `ResultSetImplementation`.

**Reply by realapril** [reply]
Drift version is 1.0.1 and drift_dev is 1.0.2, found both in pubsepc.lock file.

**Reply by realapril** [reply]
Actually both lines are problem so this table is not working too.
```
The argument type '$CategoryTable' can't be assigned to the parameter type 'ResultSetImplementation<$CategoryTable, Category>
```

**Reply by simolus3** [reply]
But if you look at the generated code, `$CategoryTable` does mixin `TableInfo`, right? Do you have a small project that reproduces this for me to look at?

**Reply by realapril** [reply]
While changing some options, my old db is broken.. How can I delete all sqlite3 db and start again? I tried with new emulator too but `Unhandled Exception: SqliteException(1): no such table , SQL logic error (code 1)` happens over and over.
I tried 
 - updating `schemaVersion` 
 - adding `android:allowBackup="false"` to manifest
 -  running app with a new emulator
 -  delete and reinstall app to the same emulator
 -  changing .sqlite file's name 
 but all were useless.

**Reply by simolus3** [reply]
If it breaks on a new emulator, it looks like the mistake happens when creating the tables. Can you enable `logStatements: true` on your `NativeDatabase`? When you start the app on a new device (or one without a database one it), does it create all the tables?
You can also use `NativeDatabase.memory()` temporarily to debug issues happening when creating a database.

If not all tables are created, do you have custom code in the `onCreate` callback of your migration? Otherwise it kind of sounds like a drift problem, but I'll likely need more information to look into it.

**Comment by KiritaniAyaka** [comment]
**TL;DR: You should use generated data class to define the field of your entry class, instead of the class written by yourself**

I have the same problem. I made a reproduction and had some try on it.

My conclusion is that **Dart gave a wrong error message**.

---

There are my classes:

```dart
class Person extends Table {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();
}

class Post extends Table {
  IntColumn get id => integer().autoIncrement()();
  IntColumn get publisher => integer().references(Person, #id)();
  TextColumn get content => text()();
}

class PostEntry {
  final Person person;
  final Post? posts;
  PostEntry(this.person, this.posts);
}
```

The same problem occurs while using the code above. The error message is also `The argument type '$PersonTable' can't be assigned to the parameter type 'ResultSetImplementation<$PersonTable, Person>'.`

As @simolus3 said, I'm exactly determined that `$PersonTable` mixin-ed `TableInfo` in generated code, so `$PersonTable` is compatible with `ResultSetImplementation` (`TableInfo` implemented `ResultSetImplementation`)

But Dart gave an error still, right? I started trying to dig with it:

Let's see the signature of the `readTable` method: 
```dart
D readTable<T extends HasResultSet, D>(ResultSetImplementation<T, D> table)
```

`tableInfo`: 
```dart
mixin TableInfo<TableDsl extends Table, D> on Table
    implements DatabaseSchemaEntity, ResultSetImplementation<TableDsl, D>
```

and generated class `$PersonTable`:
 ```dart
class $PersonTable extends Person with TableInfo<$PersonTable, PersonData>
```

Pay attention to the generic type `D`, **we can infer that `readTable` returns a `PersonData`.**

But our `PostEntry` constructor is:

```dart
class PostEntry {
  final Person person;
  final Post posts;
  PostEntry(this.person, this.posts);
}
```

Yep, the correct error message should be `The argument type 'PersonData' can't be assigned to the parameter type 'Person'`.

Changing our `PostEntry` to look like this:

```dart
class PostEntry {
  final PersonData person;
  final PostData? posts;
  PostEntry(this.person, this.posts);
}
```

The error was dismissed "magically".

---

If anybody wants to further dig with the bug of Dart, I could upload the whole project later and there are dependencies I used:

```yaml
dependencies:
  drift: ^2.6.0
  sqlite3_flutter_libs: ^0.5.13
  path_provider: ^2.0.14
dev_dependencies:
  flutter_lints: ^2.0.0
  drift_dev: ^2.6.0
  build_runner: ^2.3.3
```

```
$ flutter --version
Flutter 3.7.9 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 62bd79521d (4 days ago) • 2023-03-30 10:59:36 -0700
Engine • revision ec975089ac
Tools • Dart 2.19.6 • DevTools 2.20.1
```

**Reply by mccoyg** [reply]
@simolus3 Hi, would you please mind updating the docs: https://drift.simonbinder.eu/docs/dart-api/select/#joins - 
1. update and cap off the joins sections with an actual return for normal (two table joins) instead  of "// see next section on how to parse the result" and then never resolve it because in "Parse Results" you resolve the 2+ multiple table example.
2. Please Incorporate @KiritaniAyaka to describe how to create the data class using the generated "D" datatypes.

Even if you ignore these suggestions, Thank you!

**Reply by simolus3** [reply]
> and then never resolve it because in "Parse Results" you resolve the 2+ multiple table example.

It's actually the same example, but that's a good point - I'll change the snippet to be more complete. I've also added a comment to the `EntryWithCategory` class from the example to explicitly mention that the two fields are using the classes generated by drift instead of the source table classes.
